src/infill/SierpinskiFill.cpp

```diff
@@ -0,0 +1,802 @@
+            exchange *= -1;
+        }
+        node->error_right -= exchange;
+        next->error_left += exchange;
+    }
+
+    float total_superfluous_error = 0;
+    for (auto it = begin; it != end; ++it)
+    {
+        SierpinskiTriangle* node = *it;
+        total_superfluous_error += (distribute_subdivision_errors)? node->getSubdivisionError() : node->getValueError();
+    }
+    if (total_superfluous_error < allowed_length_error)
+    { // there is no left-over error
+        if (distribute_subdivision_errors && total_superfluous_error < -allowed_length_error)
+        {
+            std::cerr << "redistributeLeftoverErrors shouldn't be called if the node isn't to be subdivided. Total error: " << total_superfluous_error << "\n";
+            assert(false);
+        }
+        return;
+    }
+    if (begin != sequence.begin() && end != sequence.end() && first->error_left > allowed_length_error && last->error_right > allowed_length_error)
+    {
+        float total_error_input = first->error_left + last->error_right;
+        float overflow = std::min(total_superfluous_error, total_error_input);
+        float left_spillover = overflow * first->error_left / total_error_input;
+        float right_spillover = overflow * last->error_right / total_error_input;
+        (*begin)->error_left -= left_spillover;
+        prev->error_right += left_spillover;
+        (*std::prev(end))->error_right -= right_spillover;
+        next->error_left += right_spillover;
+    }
+    else if (begin != sequence.begin() && first->error_left > allowed_length_error)
+    {
+        float overflow = std::min(total_superfluous_error, first->error_left);
+        (*begin)->error_left -= overflow;
+        prev->error_right += overflow;
+        assert(first->error_left > -allowed_length_error);
+    }
+    else if (end != sequence.end() && last->error_right > allowed_length_error)
+    {
+        float overflow = std::min(total_superfluous_error, last->error_right);
+        last->error_right -= overflow;
+        next->error_left += overflow;
+        assert(last->error_right > -allowed_length_error);
+    }
+}
+
+void SierpinskiFill::balanceErrors(std::list<SierpinskiFill::SierpinskiTriangle*>::iterator begin, std::list<SierpinskiFill::SierpinskiTriangle*>::iterator end)
+{
+    // copy sublist to array
+    std::vector<SierpinskiFill::SierpinskiTriangle*> nodes;
+    for (std::list<SierpinskiFill::SierpinskiTriangle*>::iterator it = begin; it != end; ++it)
+    {
+        nodes.emplace_back(*it);
+    }
+    
+    std::vector<float> node_error_compensation(nodes.size());
+
+    // sort children on value_error, i.e. sort on total_value
+    std::vector<int> order;
+    for (unsigned int node_idx = 0; node_idx < nodes.size(); node_idx++)
+    {
+        order.emplace_back(node_idx);
+    }
+    std::sort(order.begin(), order.end(), [&nodes](int a, int b)
+        {
+            return nodes[a]->getValueError() < nodes[b]->getValueError();
+        });
```

Comment:
You have an unnecessary layer of indirection here with the node indices. You could just sort the nodes. You even already have a custom key. This would change the type of `order` from `std::vector<int>` to `std::vector<SierpinskiFill::SierpinskiTriangle*>`. You just need to alter the `sort` call to something like this:

    std::sort(order.begin(), order.end(), [](SierpinskiFill::SierpinskiTriangle* a, SierpinskiFill::SierpinskiTrianble* b)
        {
            return a->getValueError() < b->getValueError();
        });